R/all_equal_linter.R

```diff
@@ -0,0 +1,93 @@
+#' Warn about invalid usage of `all.equal()`
+#'
+#' [all.equal()] returns `TRUE` in the absence of differences but return a
+#' character string (not `FALSE`) in the presence of differences.
+#' Usage of `all.equal()` without wrapping it in `isTRUE()` in `if` clauses, or
+#' preceded by the negation operator `!`, are thus likely to generate unexpected
+#' errors if the compared objects have differences.
+#' An alternative is to use `identical()` to compare vector of strings or when
+#' exact equality is expected.
+#'
+#' @examples
+#' # lints
+#' lint(
+#'   text = 'if (all.equal(a, b)) message("equal")',
+#'   linters = all_equal_linter()
+#' )
+#'
+#' lint(
+#'   text = '!all.equal(a, b)',
+#'   linters = all_equal_linter()
+#' )
+#'
+#' lint(
+#'   text = 'isFALSE(all.equal(a, b))',
+#'   linters = all_equal_linter()
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = 'if (isTRUE(all.equal(a, b))) message("equal")',
+#'   linters = all_equal_linter()
+#' )
+#'
+#' lint(
+#'   text = '!identical(a, b)',
+#'   linters = all_equal_linter()
+#' )
+#'
+#' lint(
+#'   text = "!isTRUE(all.equal(a, b))",
+#'   linters = all_equal_linter()
+#' )
+#'
+#' @evalRd rd_tags("all_equal_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+all_equal_linter <- function() {
+
+  Linter(linter_level = "expression", function(source_expression) {
+    all_equal_calls <- source_expression$xml_find_function_calls("all.equal")
+
+    dangerous_unwrapped_all_equal <- xml_find_all(
```

Comment:
the logic is slightly off -- we get a lint here in r-devel:

https://github.com/r-devel/r-svn/blob/6d2efbc65c5ae1c15ce4f1e25e2fcf5ddc3752d2/src/library/base/R/all.equal.R#L225-L226

I guess we're not being careful about which `expr` under `IF` is being linted:

```r
if (A) all.equal(x, y)
```